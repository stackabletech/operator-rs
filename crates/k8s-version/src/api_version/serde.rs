use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::ApiVersion;

impl<'de> Deserialize<'de> for ApiVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct ApiVersionVisitor;

        impl Visitor<'_> for ApiVersionVisitor {
            type Value = ApiVersion;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                ApiVersion::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(ApiVersionVisitor)
    }
}

impl Serialize for ApiVersion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(feature = "serde")]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let _: ApiVersion =
            serde_yaml::from_str("extensions.k8s.io/v1alpha1").expect("api version is valid");
    }

    #[test]
    fn serialize() {
        let api_version =
            ApiVersion::from_str("extensions.k8s.io/v1alpha1").expect("api version is valid");
        assert_eq!(
            "extensions.k8s.io/v1alpha1\n",
            serde_yaml::to_string(&api_version).expect("api version must serialize")
        );
    }
}
