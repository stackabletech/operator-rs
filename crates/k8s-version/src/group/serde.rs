use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Group;

impl<'de> Deserialize<'de> for Group {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct GroupVisitor;

        impl<'de> Visitor<'de> for GroupVisitor {
            type Value = Group;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API group")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Group::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(GroupVisitor)
    }
}

impl Serialize for Group {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self)
    }
}
