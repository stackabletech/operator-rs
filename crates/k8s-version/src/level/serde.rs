use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Level;

impl<'de> Deserialize<'de> for Level {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct LevelVisitor;

        impl<'de> Visitor<'de> for LevelVisitor {
            type Value = Level;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version level")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Level::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(LevelVisitor)
    }
}

impl Serialize for Level {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

#[cfg(feature = "serde")]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize() {
        let _: Level = serde_yaml::from_str("alpha1").expect("level is valid");
    }

    #[test]
    fn serialize() {
        let api_version = Level::from_str("alpha1").expect("level is valid");
        assert_eq!(
            "alpha1\n",
            serde_yaml::to_string(&api_version).expect("level must serialize")
        );
    }
}
