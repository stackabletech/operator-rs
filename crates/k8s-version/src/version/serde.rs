use std::str::FromStr;

use serde::{Deserialize, Serialize, de::Visitor};

use crate::Version;

impl<'de> Deserialize<'de> for Version {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct VersionVisitor;

        impl<'de> Visitor<'de> for VersionVisitor {
            type Value = Version;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                write!(formatter, "a valid Kubernetes API version")
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Version::from_str(v).map_err(serde::de::Error::custom)
            }
        }

        deserializer.deserialize_str(VersionVisitor)
    }
}

impl Serialize for Version {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}
