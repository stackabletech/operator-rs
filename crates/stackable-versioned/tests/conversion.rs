use std::{fs::File, path::Path};

use insta::{assert_snapshot, glob};
use kube::{
    CustomResource,
    core::{conversion::ConversionReview, response::StatusSummary},
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use stackable_versioned::versioned;

#[versioned(
    k8s(group = "test.stackable.tech",),
    version(name = "v1alpha1"),
    version(name = "v1alpha2"),
    version(name = "v1beta1"),
    version(name = "v2"),
    version(name = "v3")
)]
#[derive(
    Clone,
    Debug,
    Eq,
    Hash,
    Ord,
    PartialEq,
    PartialOrd,
    CustomResource,
    Deserialize,
    JsonSchema,
    Serialize,
)]
#[serde(rename_all = "camelCase")]
struct PersonSpec {
    username: String,

    // In v1alpha2 first and last name have been added
    #[versioned(added(since = "v1alpha2"))]
    first_name: String,
    #[versioned(added(since = "v1alpha2"))]
    last_name: String,

    // We started out with a enum. As we *need* to provide a default, we have a Unknown variant.
    // Afterwards we figured let's be more flexible and accept any arbitrary String.
    #[versioned(
        added(since = "v2", default = "default_gender"),
        changed(since = "v3", from_type = "Gender")
    )]
    gender: String,
}

fn default_gender() -> Gender {
    Gender::Unknown
}

#[derive(
    Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Deserialize, JsonSchema, Serialize,
)]
#[serde(rename_all = "PascalCase")]
pub enum Gender {
    Unknown,
    Male,
    Female,
}

impl From<Gender> for String {
    fn from(value: Gender) -> Self {
        match value {
            Gender::Unknown => "Unknown".to_owned(),
            Gender::Male => "Male".to_owned(),
            Gender::Female => "Female".to_owned(),
        }
    }
}

impl From<String> for Gender {
    fn from(value: String) -> Self {
        match value.as_str() {
            "Male" => Self::Male,
            "Female" => Self::Female,
            _ => Self::Unknown,
        }
    }
}

#[test]
fn pass() {
    glob!("./inputs/conversions/pass/", "*.json", |path| {
        let (request, response) = run_for_file(path);

        let formatted = serde_json::to_string_pretty(&response)
            .expect("Failed to serialize ConversionResponse");
        assert_snapshot!(formatted);

        let response = response
            .response
            .expect("ConversionReview had no response!");

        assert_eq!(
            response.result.status,
            Some(StatusSummary::Success),
            "File {path:?} should be converted successfully"
        );
        assert_eq!(request.request.unwrap().uid, response.uid);
    })
}

#[test]
fn fail() {
    glob!("./inputs/conversions/fail/", "*.json", |path| {
        let (request, response) = run_for_file(path);

        let formatted = serde_json::to_string_pretty(&response)
            .expect("Failed to serialize ConversionResponse");
        assert_snapshot!(formatted);

        let response = response
            .response
            .expect("ConversionReview had no response!");

        assert_eq!(
            response.result.status,
            Some(StatusSummary::Failure),
            "File {path:?} should *not* be converted successfully"
        );
        if let Some(request) = &request.request {
            assert_eq!(request.uid, response.uid);
        }
    })
}

fn run_for_file(path: &Path) -> (ConversionReview, ConversionReview) {
    let request: ConversionReview =
        serde_json::from_reader(File::open(path).expect("failed to open test file"))
            .expect("failed to parse ConversionReview from test file");
    let response = Person::try_convert(request.clone());

    (request, response)
}
