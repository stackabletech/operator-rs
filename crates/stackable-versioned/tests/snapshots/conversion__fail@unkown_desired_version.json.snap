---
source: crates/stackable-versioned/tests/conversion.rs
expression: formatted
input_file: crates/stackable-versioned/tests/inputs/conversions/fail/unkown_desired_version.json
---
{
  "apiVersion": "apiextensions.k8s.io/v1",
  "kind": "ConversionReview",
  "response": {
    "uid": "c4e55572-ee1f-4e94-9097-28936985d45f",
    "result": {
      "status": "Failure",
      "code": 400,
      "message": "failed to parse desired API version: unknown API version \"test.stackable.tech/v99\"",
      "reason": "failed to parse desired API version: unknown API version \"test.stackable.tech/v99\""
    },
    "convertedObjects": []
  }
}
